### PR TITLE
(Calling) context sensitivity in the abstract interpreter

### PR DESCRIPTION
**Pull request by martin-cs**
This patch builds on #5036 and is intended as a demonstration of what the refactored infrastructure can do.  This adds optional context sensitivity to the abstract interpreter by just implementing a new history object.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Comment by martin-cs** [comment]
I guess there probably isn't too much point reviewing this in detail until #5036 is merged.  I will rebase after that.  Certainly only look at the commits that are unique to this PR.

**Comment by xbauch** [comment]
In case of reviewing: the first new commit is https://github.com/diffblue/cbmc/pull/5072/commits/76b7c1b3fc8391842ccf71acce797d6934036042.

File: src/goto-analyzer/goto_analyzer_parse_options.cpp
Comment (xbauch) [inline src/goto-analyzer/goto_analyzer_parse_options.cpp]: _an_ extended fixed-point

File: regression/goto-analyzer/command_line_01/test.desc
Comment (xbauch) [inline regression/goto-analyzer/command_line_01/test.desc:7]: Just a question: `goto-analyzer` returns `0` even when there are failing assertions?

File: src/analyses/ai_history.h
Comment (xbauch) [inline src/analyses/ai_history.h]: This could almost be a comment.

File: src/analyses/ai_history.cpp
Comment (xbauch) [inline src/analyses/ai_history.cpp:94]: Why not assert it then?

File: regression/goto-analyzer/context_sensitivity_05/main.c
Comment (xbauch) [inline regression/goto-analyzer/context_sensitivity_05/main.c:34]: I think the comment here should say success.

**Review by xbauch — APPROVED** [review]
Looks good. Could we have a test showing the difference between `domain-per-location/history`?

File: src/goto-analyzer/goto_analyzer_parse_options.cpp
Comment (hannes-steffenhagen-diffblue) [inline src/goto-analyzer/goto_analyzer_parse_options.cpp:262]: with Peter's comments on #5036 maybe this should be called `one-domain-object-per-location` or something of the sort?

File: src/goto-analyzer/goto_analyzer_parse_options.cpp
Comment (hannes-steffenhagen-diffblue) [inline src/goto-analyzer/goto_analyzer_parse_options.cpp:852]: I'd expect "ahistorical" to mean nothing is tracked, but it does track something?

File: src/analyses/ai_history.h
Comment (hannes-steffenhagen-diffblue) [reply src/analyses/ai_history.h]: This is just one of the issues with not having proper contracts support in C++ :(

File: src/analyses/ai_history.cpp
Comment (hannes-steffenhagen-diffblue) [inline src/analyses/ai_history.cpp]: `make_shared`

File: src/analyses/ai_history.cpp
Comment (hannes-steffenhagen-diffblue) [inline src/analyses/ai_history.cpp]: `make_shared`

File: src/analyses/ai_history.cpp
Comment (hannes-steffenhagen-diffblue) [inline src/analyses/ai_history.cpp:111]: Is this not an error then?

File: src/analyses/ai_history.cpp
Comment (hannes-steffenhagen-diffblue) [inline src/analyses/ai_history.cpp]: `make_shared`

File: src/analyses/ai_history.cpp
Comment (hannes-steffenhagen-diffblue) [inline src/analyses/ai_history.cpp]: `make_shared`

File: src/analyses/ai_history.cpp
Comment (hannes-steffenhagen-diffblue) [inline src/analyses/ai_history.cpp]: I'm guessing `0` here means "no limit"? I'd prefer if that was more obvious in the code, either by making `recursion_limit` optional or having a small helper like `has_recursion_limit()`

File: src/analyses/ai_history.cpp
Comment (hannes-steffenhagen-diffblue) [inline src/analyses/ai_history.cpp]: `make_shared`

File: src/analyses/ai_history.cpp
Comment (hannes-steffenhagen-diffblue) [inline src/analyses/ai_history.cpp]: Something like `number_of_recursive_calls`?

File: src/analyses/ai_history.h
Comment (hannes-steffenhagen-diffblue) [inline src/analyses/ai_history.h]: `current_stack`, `recursion_limit`

File: src/analyses/ai_history.cpp
Comment (hannes-steffenhagen-diffblue) [inline src/analyses/ai_history.cpp:125]: `make_shared`

File: src/analyses/ai_history.h
Comment (hannes-steffenhagen-diffblue) [inline src/analyses/ai_history.h]: `make_shared`

File: src/analyses/ai_history.h
Comment (hannes-steffenhagen-diffblue) [inline src/analyses/ai_history.h]: `make_shared`

File: src/goto-analyzer/static_verifier.cpp
Comment (hannes-steffenhagen-diffblue) [inline src/goto-analyzer/static_verifier.cpp]: I agree this seems to be an improvement, but it appears unrelated to the rest of the PR?

File: src/goto-analyzer/static_verifier.cpp
Comment (hannes-steffenhagen-diffblue) [inline src/goto-analyzer/static_verifier.cpp:287]: I believe the `UNREACHABLE` is unnecessary as this place is just syntactically unreachable right now

File: src/goto-analyzer/static_verifier.cpp
Comment (hannes-steffenhagen-diffblue) [inline src/goto-analyzer/static_verifier.cpp:441]: So normally `FALSE` just meant "definitely fails and is definitely reachable" IIRC? Isn't this breaking convention?

File: regression/goto-analyzer/context_sensitivity_05/main.c
Comment (hannes-steffenhagen-diffblue) [reply regression/goto-analyzer/context_sensitivity_05/main.c:34]: In context this would always be true, but I believe this is about the semantics of `--constants` where it can also be unknown - however, this definitely can't be `false`.

**Review by hannes-steffenhagen-diffblue — COMMENTED** [review]
Looks good to me generally, just some questions/nitpicks

File: regression/goto-analyzer/command_line_01/test.desc
Comment (martin-cs) [reply regression/goto-analyzer/command_line_01/test.desc:7]: Yeah... At the moment it does because the analysis has run successfully / it hasn't shown the presence of bugs.  I can see an argument for it returning 10 but I think that is beyond the scope of the current PR.

File: src/analyses/ai_history.h
Comment (martin-cs) [reply src/analyses/ai_history.h]: I have seen contract languages for C and C++ and they get kinda scary kinda quickly.

File: src/analyses/ai_history.cpp
Comment (martin-cs) [reply src/analyses/ai_history.cpp:94]: I did but there is a legitimate case when this happens -- the end of the initial calling function.  This is the code that finally finishes the analysis.

File: src/goto-analyzer/goto_analyzer_parse_options.cpp
Comment (martin-cs) [reply src/goto-analyzer/goto_analyzer_parse_options.cpp:852]: The object knows it's current location but nothing more; it knows nothing of it's history.  I mean, ahistorical is kinda a bad name but I couldn't come up with anything more exact.

File: src/analyses/ai_history.cpp
Comment (martin-cs) [reply src/analyses/ai_history.cpp:111]: It's not an error.  The `ait` dispatch loop doesn't know which transitions are possible so it simply tries all of them (in this case, trying all transitions from *any* history that reaches the end of the callee function back to the call site) and relies on the history to block out any of these that the history can rule impossible.

This is kind of subtle and the first time this feature has been used; that's sort of why I wanted to write this patch to demonstrate this.

File: src/goto-analyzer/static_verifier.cpp
Comment (martin-cs) [reply src/goto-analyzer/static_verifier.cpp]: Sort of.  The next commit refactors the same bit of code so I wanted to separate the 'improvement that changes the regression tests a bit' from the 'refactoring that adds new capabilities'.

File: src/goto-analyzer/static_verifier.cpp
Comment (martin-cs) [reply src/goto-analyzer/static_verifier.cpp:287]: It is right now, I tend to use them defensively so I can't later break the code by removing an option / adding a value to the enum without finding out.

File: src/goto-analyzer/static_verifier.cpp
Comment (martin-cs) [reply src/goto-analyzer/static_verifier.cpp:441]: It is not really possible or meaningful to follow BMC style conventions in an abstract interpreter.  In BMC we are always dealing with the possibility that there are executions that we haven't explored, so "TRUE / SAFE" comes with an implicit "(in all the traces we have explored)" but on the plus side "FALSE / UNSAFE" is a definite thing.

In the AI case, "TRUE / SAFE" really is in *all* traces but "FALSE / UNSAFE (if reachable)" is the best we can manage.

I'm not sure I see an easy way of resolving this; it is certainly more fundamental than this PR.

**Comment by martin-cs** [comment]
My thanks to @xbauch and @hannes-steffenhagen-diffblue for their comments.  I have responded to all of the 'discussion' ones.  I may leave the 'changes required' ones until the rebase once the main PR is merged.

**Review by hannes-steffenhagen-diffblue — APPROVED** [review]
Going over my comments and your responses, unless there are substantial changes after the rebase this'll probably not require a re-review from me. (I think the `make_shared`s are needed for consistency and _some_ correctness though)

File: regression/goto-analyzer/command_line_02/main.c
Comment (chrisr-diffblue) [inline regression/goto-analyzer/command_line_02/main.c:1]: Since this test file appears to be identical to the one in command_line_01, why not just have a single file, and use two `.desc` files in the one directory?

File: src/analyses/ai_history.cpp
Comment (chrisr-diffblue) [reply src/analyses/ai_history.cpp:111]: Probably out of scope of this PR, but how would all this work with exceptions or something like setjmp/longjmp ?

File: src/analyses/ai_history.h
Comment (chrisr-diffblue) [inline src/analyses/ai_history.h:301]: Rather than just a comment saying that the default implementation is used, I still prefer to explicitly define the function and invoke the parent.

**Review by chrisr-diffblue — APPROVED** [review]
Some minor queries and nit picks, but nothing blocking.

File: src/analyses/ai_history.cpp
Comment (martin-cs) [reply src/analyses/ai_history.cpp:111]: At the moment, it doesn't.  `ait` will through an exception if it finds try `THROW` or `CATCH` instructions and some domains will also fail in transform if they meet them.  So the following is hypothetical but...

(-2. Find a use case for analysing setjmp/longjmp. )
(-1. Get setjmp/longjmp working with the rest of the code. )
0. Actually get some agreement and documentation of the semantics of THROW/CATCH.
1. Inherit from `ait` and create an abstract interpreter that doesn't use recursion for interprocedural stuff.  This will need to track call-sites for functions.
2. Add tracking of catch locations and override `visit` to handle throw and catch.  Throw will then attempt to step edges to all possible catch locations.
3. Modify this, or another history so that it blocks steps to implausible exception handlers (i.e. not in the call stack) as given above.
4. Modify domains as needed so they can compute the edge.

Thinking about this ... actually the part from 1 to 4 is probably a single PR's worth.  I suspect the hard work is up to and including 0.

**Comment by martin-cs** [comment]
TODO : Triggering the recursion limit should result in widening returning true.

**Comment by martin-cs** [comment]
@chrisr-diffblue : exceptions and `setjmp`/`longjmp`, possible but beyond the scope of this PR, way beyond.

Somewhere I have a written up plan for how to do exceptions (assuming anyone can document the semantics of CPROVER exceptions) but the short version is to compute an "edge" from every throw to every (previously reached) catch in `ai_baset`.  The `step` function in the history allows these to be pruned down to only those that are historically feasible -- that would work really well for this calling context history.

`setjmp`/`longjmp` ... is an interesting challenge ... that may have to wait until someone has some real code using it that they want to analyse.  In principle you would think that you could associate the `jmp_buf` object with a set of histories and then on `longjmp` you could compute "edges" to each of these and again, let `step` prune them.  BUT that wouldn't be enough as you would also have to overwrite the globals and heap variables in the saved domains with those from the caller and there isn't currently a good interface for doing that and it might require changes to the way domains are handled... yeah... not a starter task...

File: src/goto-analyzer/goto_analyzer_parse_options.cpp
Comment (martin-cs) [reply src/goto-analyzer/goto_analyzer_parse_options.cpp]: Fixed.

File: regression/goto-analyzer/context_sensitivity_05/main.c
Comment (martin-cs) [reply regression/goto-analyzer/context_sensitivity_05/main.c:34]: I think they should say true and false because the comments are about the truth of the propositions.  To me success and failure are about the tool being able to / not being able to prove something.

File: src/goto-analyzer/goto_analyzer_parse_options.cpp
Comment (martin-cs) [reply src/goto-analyzer/goto_analyzer_parse_options.cpp:262]: See https://github.com/diffblue/cbmc/issues/5197 for discussion of this kind of thing.

File: src/analyses/ai_history.cpp
Comment (martin-cs) [reply src/analyses/ai_history.cpp]: Fixed.

File: src/analyses/ai_history.cpp
Comment (martin-cs) [reply src/analyses/ai_history.cpp]: Fixed.

File: src/analyses/ai_history.cpp
Comment (martin-cs) [reply src/analyses/ai_history.cpp]: Fixed.

File: src/analyses/ai_history.cpp
Comment (martin-cs) [reply src/analyses/ai_history.cpp]: Fixed.

File: src/analyses/ai_history.cpp
Comment (martin-cs) [reply src/analyses/ai_history.cpp]: Fixed.

File: src/analyses/ai_history.cpp
Comment (martin-cs) [reply src/analyses/ai_history.cpp:125]: Fixed.

File: src/analyses/ai_history.h
Comment (martin-cs) [reply src/analyses/ai_history.h]: Fixed.

File: src/analyses/ai_history.h
Comment (martin-cs) [reply src/analyses/ai_history.h]: Fixed.

File: src/analyses/ai_history.cpp
Comment (martin-cs) [reply src/analyses/ai_history.cpp:125]: Or... potentially not fixed.  Turns out that `std::make_shared` doesn't work with protected and private constructors.  All of the work-arounds are more horrible than leaving this one instance:

https://stackoverflow.com/questions/54454620/why-cant-i-access-the-private-constructor-in-my-static-function-as-below/54456533#54456533
https://stackoverflow.com/questions/3541632/using-make-shared-with-a-protected-constructor-abstract-interface

( Yes, `friend` should work, no, it doesn't. )

I've had a chat with @hannes-steffenhagen-diffblue about this and he says that leaving the new is the least worst option.

File: src/analyses/ai_history.cpp
Comment (martin-cs) [reply src/analyses/ai_history.cpp]: Done.

File: src/analyses/ai_history.cpp
Comment (martin-cs) [reply src/analyses/ai_history.cpp]: Done.

File: src/analyses/ai_history.h
Comment (martin-cs) [reply src/analyses/ai_history.h]: Renamed to `cur_stack` and `rec_lim` so the naming doesn't clash with member variables.

File: regression/goto-analyzer/command_line_02/main.c
Comment (martin-cs) [reply regression/goto-analyzer/command_line_02/main.c:1]: Fair point, will do in future.  Don't particularly want to change this right now as a hassle getting this back into the right commit.

File: src/analyses/ai_history.h
Comment (martin-cs) [reply src/analyses/ai_history.h:301]: Done!

**Comment by martin-cs** [comment]
@xbauch , @hannes-steffenhagen-diffblue and @chrisr-diffblue : I have made the requested changes and have pushed a rebased version ( post merging #5036 ).  If^WWhen this passes CI I will squash back as many of the fix commits as possible and then merge.  Shout if you want something else done!

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: ae0d7ade97d9437d0d88ae61738257751fba7809).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/138337121

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: fddd0987118614033edb7c90ccacb8c4b5df1ebb).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/138404904

**Pull request merged**
